`bom.vulnerabilities` model

thepwagner opened the pull request:
Add [Vulnerabilties](https://cyclonedx.org/docs/1.4/json/#vulnerabilities) to the Bom model, including enums as needed.

The implementation is styled after the existing `Component` implementation and tests.

this is part of #164

jkowalleck commented:
Thank you for this feature.  Review will take me a while. 
Are you planning on implementing the JSON- and XML normalizer, too?

jkowalleck commented on src/enums/vulnerabilityRatingSeverity.ts:
```diff
@@ -0,0 +1,28 @@
+Copyright (c) OWASP Foundation. All Rights Reserved.
+*/
+
+export enum VulnerabiltyRatingSeverity {
```
:x: please add funcional tests ala https://github.com/CycloneDX/cyclonedx-javascript-library/blob/main/tests/functional/Enums.ComponentScope.spec.js

jkowalleck commented on src/enums/vulnerabilityRatingMethod.ts:
```diff
@@ -0,0 +1,26 @@
+Copyright (c) OWASP Foundation. All Rights Reserved.
+*/
+
+export enum VulnerabilityRatingMethod {
```
:x: please add functional tests ala https://github.com/CycloneDX/cyclonedx-javascript-library/blob/main/tests/functional/Enums.ComponentScope.spec.js

jkowalleck commented on src/enums/vulnerabilityAnalysisState.ts:
```diff
@@ -0,0 +1,27 @@
+Copyright (c) OWASP Foundation. All Rights Reserved.
+*/
+
+export enum VulnerabilityAnalysisState {
```
:x: please add functional tests ala https://github.com/CycloneDX/cyclonedx-javascript-library/blob/main/tests/functional/Enums.ComponentScope.spec.js

jkowalleck commented on src/enums/vulnerabilityAnalysisResponse.ts:
```diff
@@ -0,0 +1,26 @@
+Copyright (c) OWASP Foundation. All Rights Reserved.
+*/
+
+export enum VulnerabilityAnalysisResponse {
```
:x: please add functional tests ala https://github.com/CycloneDX/cyclonedx-javascript-library/blob/main/tests/functional/Enums.ComponentScope.spec.js

jkowalleck commented on src/enums/vulnerabilityAnalysisJustification.ts:
```diff
@@ -0,0 +1,30 @@
+Copyright (c) OWASP Foundation. All Rights Reserved.
+*/
+
+export enum VulnerabilityAnalysisJustification {
```
:x: please add functional tests ala https://github.com/CycloneDX/cyclonedx-javascript-library/blob/main/tests/functional/Enums.ComponentScope.spec.js

jkowalleck commented on src/enums/vulnerabilityAffectsVersionStatus.ts:
```diff
@@ -0,0 +1,24 @@
+Copyright (c) OWASP Foundation. All Rights Reserved.
+*/
+
+export enum VulnerabilityAffectsVersionStatus {
```
:x: please add functional tests ala https://github.com/CycloneDX/cyclonedx-javascript-library/blob/main/tests/functional/Enums.ComponentScope.spec.js

jkowalleck replied:
rename to `ImpactAnalysisAffectedStatusType` like in XML schema definition? 
rename to `AffectedStatus` like in JSON schema definition?

jkowalleck commented on tests/unit/Models.Vulnerability.spec.js:
```diff
@@ -0,0 +1,140 @@
+    assert.strictEqual(vulnerability.bomRef.value, undefined)
+    assert.strictEqual(vulnerability.id, undefined)
+    assert.strictEqual(vulnerability.source, undefined)
+    assert.strictEqual(vulnerability.references.size, 0)
```
:x: please add type checks -- as a regression test to find unintended type chages (which would lead to breaking changes)

jkowalleck commented on tests/unit/Models.Vulnerability.spec.js:
```diff
@@ -0,0 +1,140 @@
+    assert.strictEqual(vulnerability.properties.size, 0)
+  })
+
+  test('constructor with OptionalProperties', () => {
```
:+1:

jkowalleck reviewed (changes requested):
partial review. 

please find critical remarks with an :x: .
other remarks are discussion items.

thepwagner commented on src/enums/vulnerabilityAnalysisResponse.ts:
```diff
@@ -0,0 +1,26 @@
+Copyright (c) OWASP Foundation. All Rights Reserved.
+*/
+
+export enum VulnerabilityAnalysisResponse {
```
I switched to `ImpactAnalysisResponse` as this was the most consistent with the JSON schema, even if the JSON schema doesn't give this enum a standalone definition.

Please advise if you'd prefer a different name.

thepwagner commented on src/models/vulnerability.ts:
```diff
@@ -0,0 +1,110 @@
+  source?: Source
+  references: VulnerabilityReferenceRepository
+  ratings: VulnerabilityRatingRepository
+  cwes: number[]
```
Thanks for this tip! Reusing `PositiveInteger` made this easy. Do you think `CWE[]` is the best type for this collection?


Something like `SortableSet<CWE>` would be great, but I couldn't sort out the TypeScript to have the `CWE` type implement `Comparable`.

thepwagner commented:
@jkowalleck thank you for the quick initial feedback!
I've addressed feedback and resolved the threads I think are resolved - please advise if you'd prefer to resolve after verifying the fixes.

* I made all suggested renames, biased towards the naming of the JSON schema.
* I've extended the `compare()` methods to consider all fields, including nested sets like `Set<URL | string>` or `SortableSet<VulnerabilityReference>`. This meant adding a generic `SortableSet.compare()`
* Using the JSON schema for functional tests is cool! I extended the name-mapping helper to support `snake_case` since there was no feedback on the naming of enum values in the first review.

I was not intending to implement the normalizers.

jkowalleck commented on src/models/vulnerabilityCredits.ts:
```diff
@@ -0,0 +1,50 @@
+}
+
+export class VulnerabilityCredits implements Comparable {
+  organizations?: OrganizationalEntityRepository
```
:x: repositories must not be optiional, but they may be empty

jkowalleck commented on src/models/vulnerabilityCredits.ts:
```diff
@@ -0,0 +1,50 @@
+
+export class VulnerabilityCredits implements Comparable {
+  organizations?: OrganizationalEntityRepository
+  individuals?: OrganizationalContactRepository
```
:x: repositories must not be optiional, but they may be empty

jkowalleck commented on src/models/vulnerabilityRating.ts:
```diff
@@ -0,0 +1,56 @@
+  }
+
+  compare (other: VulnerabilityRating): number {
+    return (this.score ?? 0) - (other.score ?? 0)
```
this appears to be like an insufficient comparison.
if this is enough to properly compare, then please add a in-code comment that explains why.

jkowalleck commented on src/models/vulnerabilityReference.ts:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  compare (other: VulnerabilityReference): number {
+    return (this.id ?? '').localeCompare(other.id ?? '')
```
:x:  unnecessary null-comparisson

jkowalleck commented on src/models/vulnerabilitySource.ts:
```diff
@@ -0,0 +1,44 @@
+import { Comparable, SortableSet } from '../helpers/sortableSet'
+
+interface OptionalProperties {
+  url?: URL | string
```
:x:  use type refenrences ala `VulnerabilitySource['url']`

jkowalleck reviewed (changes requested):
please find critical chaneg requests marked with :x: '
everything else is a discussion item.

jkowalleck commented:
thank you for all your work, @thepwagner 

could you open a pull request with your latest things and have it target the "feature/vulnerabilities" branch ? 
Plan is to merge your models in there, fine tune them a bit and merge them into master for an early release.

Then, in a later feature, the normalizers and serializers can be developed and be released separately. 

This way you (ad others) will have the models for use soon,
and will not be blocked by normalization details.

jkowalleck commented:
work continues in #419
